client/src/procedures/primitives.rs

```diff
@@ -0,0 +1,1008 @@
+    Sha2_256(Hash<Sha256>),
+    Sha2_384(Hash<Sha384>),
+    Sha2_512(Hash<Sha512>),
+    Blake2b256(Hash<Blake2b256>),
+}
+
+impl ProcedureStep for Hashes {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error> {
+        match self {
+            Hashes::Sha2_256(proc) => proc.execute(runner, state),
+            Hashes::Sha2_384(proc) => proc.execute(runner, state),
+            Hashes::Sha2_512(proc) => proc.execute(runner, state),
+            Hashes::Blake2b256(proc) => proc.execute(runner, state),
+        }
+    }
+}
+
+#[derive(Procedure, Clone, Serialize, Deserialize)]
+pub struct Hash<T> {
+    #[input_data]
+    msg: InputData<Vec<u8>>,
+
+    #[output_key]
+    output_key: InterimProduct<OutputKey>,
+
+    _marker: PhantomData<T>,
+}
+
+impl<T> Hash<T> {
+    pub fn new(msg: Vec<u8>) -> Self {
+        Hash {
+            msg: InputData::Value(msg),
+            output_key: InterimProduct {
+                target: OutputKey::random(),
+                is_temp: true,
+            },
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn dynamic(msg_key: OutputKey) -> Self {
+        let input = InputData::Key(msg_key);
+        Hash {
+            msg: input,
+            output_key: InterimProduct {
+                target: OutputKey::random(),
+                is_temp: true,
+            },
+            _marker: PhantomData,
+        }
+    }
+}
+
+#[execute_procedure]
+impl<T: Digest> Parse for Hash<T> {
+    type Input = Vec<u8>;
+    type Output = Vec<u8>;
+
+    fn parse(self, input: Self::Input) -> Result<Self::Output, engine::Error> {
+        let mut digest = vec![0; T::OutputSize::USIZE];
+        digest.copy_from_slice(&T::digest(&input));
+        Ok(digest)
+    }
+}
+
+#[derive(Procedure, Clone, Serialize, Deserialize)]
+pub enum Hmacs {
+    Sha2_256(Hmac<Sha256>),
+    Sha2_384(Hmac<Sha384>),
+    Sha2_512(Hmac<Sha512>),
+}
+
+impl ProcedureStep for Hmacs {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error> {
+        match self {
+            Hmacs::Sha2_256(proc) => proc.execute(runner, state),
+            Hmacs::Sha2_384(proc) => proc.execute(runner, state),
+            Hmacs::Sha2_512(proc) => proc.execute(runner, state),
+        }
+    }
+}
+
+#[derive(Procedure, Clone, Serialize, Deserialize)]
+pub struct Hmac<T> {
+    #[input_data]
+    msg: InputData<Vec<u8>>,
+
+    #[output_key]
+    output_key: InterimProduct<OutputKey>,
+
+    #[source]
+    key: Location,
+
+    _marker: PhantomData<T>,
+}
+
+impl<T> Hmac<T> {
+    pub fn new(msg: Vec<u8>, key: Location) -> Self {
+        Hmac {
+            msg: InputData::Value(msg),
+            key,
+            output_key: InterimProduct {
+                target: OutputKey::random(),
+                is_temp: true,
+            },
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn dynamic(msg_key: OutputKey, key: Location) -> Self {
```

Comment:
the functions `new` and `dynamic` (with similar intend) appear more than once. Might it be possible to combine them into a trait?

Reply:
Generally, I think is a good suggestion and would be desirable to have such a trait. The problem is that the different procedures have different number of input arguments, so adding a trait would require in some cases to wrap the other input parameters  in a new struct, so we can have the signatures:
```
fn new(input: T, others: U) -> Self;
fn dynamic(msg_key:  OutputKey, other: U) -> Self
```
with `T` being the input type, and U  being other parameters.